cli/data.go

```diff
@@ -636,120 +699,151 @@ func filenameForDownload(meta *datapb.BinaryMetadata) string {
-			m["MetadataIndex"] = localToGlobalMDIndex[int(datum.GetMetadataIndex())]
-			j, err := json.Marshal(m)
-			if err != nil {
-				return errors.Wrap(err, "could not marshal JSON response")
-			}
-			_, err = w.Write(append(j, []byte("\n")...))
-			if err != nil {
-				return errors.Wrapf(err, "could not write to file %s", dataFile.Name())
+			var exportErr error
+
+			ctx, cancel := context.WithCancel(context.Background())
+
+			defer func() {
+				writer.Flush()   //nolint:errcheck,gosec
+				dataFile.Close() //nolint:errcheck,gosec
+				cancel()
+
+				if exportErr != nil {
+					os.Remove(dataFile.Name()) //nolint:errcheck,gosec
+				}
+			}()
+
+			go func() {
+				defer close(dataRowChan)
+				fmt.Fprintf(c.c.App.Writer, ".") //nolint:errcheck // Adds '.' to 'Downloading..' output.
+
+				stream, err := c.dataClient.ExportTabularData(ctx, request)
+				if err != nil {
+					errChan <- errors.Wrap(err, "failed to export tabular data")
+					return
+				}
+
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+						resp, err := stream.Recv()
+						if errors.Is(err, io.EOF) {
+							return
+						}
+						if err != nil {
+							errChan <- errors.Wrap(err, "error receiving tabular data")
+							return
+						}
+
+						dataRow, err := protojson.Marshal(resp)
+						if err != nil {
+							errChan <- errors.Wrap(err, "error formatting tabular data")
+							return
+						}
+
+						select {
+						case dataRowChan <- dataRow:
+							// Successfully sent.
+						case <-ctx.Done():
+							return
+						}
+					}
+				}
+			}()
+
+			for {
+				select {
+				case dataRow, ok := <-dataRowChan:
+					// No more data to write.
+					if !ok {
+						if err = writer.Flush(); err != nil {
+							exportErr = errors.Wrap(err, "error writing data to file")
+							return exportErr
+						}
+
+						return nil
+					}
+
+					if err = writeData(writer, dataRow, &numWrites); err != nil {
+						exportErr = errors.Wrap(err, "error writing data")
+						return exportErr
+					}
+				case err := <-errChan:
+					exportErr = err
+					return err
+				case <-ctx.Done():
+					exportErr = ctx.Err()
+					return ctx.Err()
+				}
 			}
+		}()
+
+		if err != nil && count < maxRetryCount-1 {
+			continue
 		}
-	}
 
-	printf(c.c.App.Writer, "") // newline
-	if err := w.Flush(); err != nil {
-		return errors.Wrapf(err, "could not flush writer for %s", dataFile.Name())
+		printf(c.c.App.Writer, "") // newline
+		return err
 	}
 
 	return nil
 }
 
-func makeDestinationDirs(dst string) error {
-	if err := os.MkdirAll(filepath.Join(dst, dataDir), 0o700); err != nil {
+func writeData(writer *bufio.Writer, dataRow []byte, numWrites *uint64) error {
+	*numWrites++
+
+	_, err := writer.Write(dataRow)
+	if err != nil {
+		return err
+	}
+
+	err = writer.WriteByte('\n')
+	if err != nil {
 		return err
 	}
-	if err := os.MkdirAll(filepath.Join(dst, metadataDir), 0o700); err != nil {
+
+	// Periodically flush to keep buffer size down.
```

Comment:
instead of counting writes (which can very in size), how about tracking buffer size instead? you get the number of bytes written from `writer.Write` and can add that instead of `++`